metadata-ingestion/tests/unit/test_bigquery_source.py

```diff
@@ -405,3 +406,101 @@ def test_table_processing_logic_date_named_tables(client_mock, data_dictionary_m
     ]  # alternatively
     for table in tables.keys():
         assert tables[table].table_id in ["test-table", "20220103"]
+
+
+@pytest.mark.parametrize(
+    "table_name, expected_table_prefix, expected_shard",
+    [
+        # Cases with Fully qualified name as input
+        ("project.dataset.table", "project.dataset.table", None),
+        ("project.dataset.table_20231215", "project.dataset.table", "20231215"),
+        ("project.dataset.table_2023", "project.dataset.table_2023", None),
+        # incorrectly handled special case where dataset itself is a sharded table if full name is specified
+        ("project.dataset.20231215", "project.dataset.20231215", None),
+        # Cases with Just the table name as input
+        ("table", "table", None),
+        ("table20231215", "table20231215", None),
+        ("table_20231215", "table", "20231215"),
+        ("table_1624046611000_name", "table_1624046611000_name", None),
+        ("table_1624046611000", "table_1624046611000", None),
+        # Special case where dataset itself is a sharded table
+        ("20231215", None, "20231215"),
+    ],
+)
```

Comment:
What do you guys think about making these doctests instead? Seems like a standard use of them, as the tests are simple and help document the function

Reply:
I like the idea, but won't that require some framework level update to execute them ?

Reply:
Hmm didn't realize we didn't have any. Definitely a separate change then, I'll try looking into it

Reply:
Yep we don't have doctests yet, but we can enable them by adding `--doctest-modules` in pytest to the setup.cfg file

Definitely agree that they'd help in these cases